src/DomainHost.sol

```diff
@@ -0,0 +1,411 @@
+    }
+
+    /// @notice Initiate shutdown for this domain
+    /// @dev This will trigger the end module on the remote domain
+    function _cage() internal returns (bytes memory payload) {
+        require(vat.live() == 0 || wards[msg.sender] == 1, "DomainHost/not-authorized");
+        require(live == 1, "DomainHost/not-live");
+
+        live = 0;
+
+        payload = abi.encodeWithSelector(DomainGuest.cage.selector, rid++);
+
+        emit Cage();
+    }
+
+    /// @notice Set this domain's cure value
+    /// @param _lid Local ordering id
+    /// @param value The value of the cure [RAD]
+    function tell(uint256 _lid, uint256 value) external guestOnly ordered(_lid) {
+        require(live == 0, "DomainHost/live");
+        require(!cureReported, "DomainHost/cure-reported");
+        require(_divup(value, RAY) <= grain, "DomainHost/cure-bad-value");
+
+        cureReported = true;
+        cure = value;
+
+        emit Tell(value);
+    }
+
+    /// @notice Allow DAI holders to exit during global settlement
+    /// @dev    This will mint a pro-rata claim token on the remote domain.
+    ///         Gem amount is scaled by the actual debt of the remote domain.
+    ///         `usr` is the address for the mint on the remote domain.
+    /// @param usr The address to send the claim token to
+    /// @param wad The amount of gems to exit [WAD]
+    function _exit(address usr, uint256 wad) internal returns (bytes memory payload) {
+        require(vat.live() == 0, "DomainHost/vat-live");
+        vat.slip(ilk, msg.sender, -_int256(wad));
+
+        // Convert to actual debt amount
+        // Round against the user
+        uint256 claimAmount = wad * (grain - _divup(cure, RAY)) / grain;
+        
+        payload = abi.encodeWithSelector(DomainGuest.exit.selector, usr, claimAmount);
+
+        emit Exit(usr, wad, claimAmount);
+    }
+
+    /// @notice Undo an exit
+    /// @dev    Some chains do not guarantee inclusion of a transaction.
+    ///         This function allows the user to undo an exit if it is not relayed
+    ///         to the other side.
+    ///         It is up to the implementation to ensure this message was not relayed
+    ///         otherwise you open yourself up to double spends.
+    /// @param originalSender The msg.sender from the _exit() call
+    /// @param wad The amount of gems that was attempted to exit [WAD]
+    function _undoExit(address originalSender, uint256 wad) internal {
+        vat.slip(ilk, originalSender, _int256(wad));
+
+        emit UndoExit(originalSender, wad);
+    }
+
+    // --- Canonical DAI Support ---
+
+    /// @notice Deposit local DAI to mint remote canonical DAI
+    /// @param to The address to send the DAI to on the remote domain
+    /// @param amount The amount of DAI to deposit [WAD]
+    function _deposit(address to, uint256 amount) internal returns (bytes memory payload) {
+        require(dai.transferFrom(msg.sender, escrow, amount), "DomainHost/transfer-failed");
```

Comment:
I know it won't happen from a practical point of view but I'd feel better from the correctness one to put the require that `amount` can't be higher than `int256` which would break the guest part.

Reply:
Actually thinking on this. Do we want this practically useless check? This may be a fairly common path and so gas consumption is important.

Reply:
You are probably right it is not worth it.